jsonschema/src/schema_node.rs

```diff
@@ -0,0 +1,429 @@
+        unmatched_keywords: Option<AHashMap<String, serde_json::Value>>,
+    ) -> SchemaNode {
+        validators.shrink_to_fit();
+        SchemaNode {
+            relative_path: context.clone().into_pointer(),
+            absolute_path: context.base_uri().map(AbsolutePath::from),
+            validators: NodeValidators::Keyword(Box::new(KeywordValidators {
+                unmatched_keywords,
+                validators,
+            })),
+        }
+    }
+
+    pub(crate) fn new_from_array(
+        context: &CompilationContext<'_>,
+        mut validators: Vec<BoxedValidator>,
+    ) -> SchemaNode {
+        validators.shrink_to_fit();
+        SchemaNode {
+            relative_path: context.clone().into_pointer(),
+            absolute_path: context.base_uri().map(AbsolutePath::from),
+            validators: NodeValidators::Array { validators },
+        }
+    }
+
+    pub(crate) fn validators(&self) -> impl Iterator<Item = &BoxedValidator> + ExactSizeIterator {
+        match &self.validators {
+            NodeValidators::Boolean { validator } => {
+                if let Some(v) = validator {
+                    NodeValidatorsIter::BooleanValidators(std::iter::once(v))
+                } else {
+                    NodeValidatorsIter::NoValidator
+                }
+            }
+            NodeValidators::Keyword(kvals) => {
+                NodeValidatorsIter::KeywordValidators(kvals.validators.iter())
+            }
+            NodeValidators::Array { validators } => {
+                NodeValidatorsIter::ArrayValidators(validators.iter())
+            }
+        }
+    }
+
+    fn format_validators(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", format_validators(self.validators()))
+    }
+
+    /// This is simmilar to `Validator::apply` except that `SchemaNode` knows where it is in the
+    /// validator tree and so rather than returning a `PartialApplication` it is able to return a
+    /// complete `BasicOutput`. This is the mechanism which compositional validators use to combine
+    /// results from sub-schemas
+    pub(crate) fn apply_rooted(
+        &self,
+        schema: &JSONSchema,
+        instance: &serde_json::Value,
+        instance_path: &InstancePath,
+    ) -> BasicOutput {
+        match self.apply(schema, instance, instance_path) {
+            PartialApplication::Valid {
+                annotations,
+                mut child_results,
+            } => {
+                if let Some(annotations) = annotations {
+                    child_results.insert(0, self.annotation_at(instance_path, annotations));
+                };
+                BasicOutput::Valid(child_results)
+            }
+            PartialApplication::Invalid {
+                errors,
+                mut child_results,
+            } => {
+                for error in errors {
+                    child_results.insert(0, self.error_at(instance_path, error));
```

Comment:
Would it make sense to use `VecDeque` here instead?  to avoid moving elements

Reply:
Good shout, fixed.